chainstate/tx-verifier/src/transaction_verifier/input_output_policy.rs

```diff
@@ -13,75 +13,773 @@
+                                }
+                                TxOutput::ProduceBlockFromStake(_, _, _) => Ok(()),
+                            },
+                            _ => Err(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::MultipleBlockRewardOutputs,
+                            )),
+                        }
+                    }
+                },
+                // multiple inputs
+                _ => Err(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::ConsensusPoSError(ConsensusPoSError::MultipleKernels),
+                )),
+            }
+        }
+        None => {
+            // if no kernel input is present it's allowed to have multiple `LockThenTransfer` outputs
+            let all_lock_then_transfer = reward
+                .outputs()
+                .ok_or(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::NoBlockRewardOutputs,
+                ))?
+                .iter()
+                .all(|output| match output {
+                    TxOutput::LockThenTransfer(_, _, _) => true,
+                    TxOutput::Transfer(_, _)
+                    | TxOutput::Burn(_)
+                    | TxOutput::StakePool(_)
+                    | TxOutput::ProduceBlockFromStake(_, _, _)
+                    | TxOutput::DecommissionPool(_, _, _, _) => false,
+                });
+            ensure!(
+                all_lock_then_transfer,
+                ConnectTransactionError::InvalidOutputTypeInReward
+            );
+            Ok(())
         }
     }
 }
 
-/// Indicates whether an output purpose can be used in a tx as an output
-fn is_valid_output_for_tx(output: &TxOutput) -> bool {
-    match output {
-        TxOutput::Transfer(_, _)
-        | TxOutput::LockThenTransfer(_, _, _)
-        | TxOutput::Burn(_)
-        | TxOutput::StakePool(_) => true,
-        TxOutput::ProduceBlockFromStake(_, _, _) => false,
-    }
-}
-
-fn check_inputs_can_be_used_in_tx(
+/// Not all `TxOutput` combinations can be used in a transaction.
+pub fn check_tx_inputs_outputs_purposes(
     tx: &Transaction,
     utxo_view: &impl utxo::UtxosView,
 ) -> Result<(), ConnectTransactionError> {
-    let can_be_spent = tx
+    let inputs = tx
         .inputs()
         .iter()
         .map(|input| {
             utxo_view
                 .utxo(input.outpoint())
                 .map_err(|_| utxo::Error::ViewRead)?
+                .map(|u| u.output().clone())
                 .ok_or(ConnectTransactionError::MissingOutputOrSpent)
         })
-        .collect::<Result<Vec<_>, _>>()?
-        .iter()
-        .all(|utxo| is_valid_input_for_tx(utxo.output()));
+        .collect::<Result<Vec<_>, _>>()?;
+
+    match inputs.as_slice() {
+        // no inputs
+        [] => return Err(ConnectTransactionError::MissingTxInputs),
+        // single input
+        [input] => match tx.outputs() {
```

Comment:
What I don't understand here is... what's the value of distinguishing between one input or many? There's no special circumstance where one input is the right way to do things (except for staking). In fact, most of the time the number of inputs is irrelevant.

But then let's assume this is the right way to do it... this just means there has to be (by definition), duplicate code, which is what I'm seeing at the bottom of both arms of one or more inputs... well, why aren't these taking outside to a private function?

Reply:
The rules for one and many are different. For example, it's ok to have `StakePool` -> `DecommissionPool`, but it's not ok to have `StakePool` + `Transfer` -> `DecommissionPool`.
So I come up with a solution to explicitly distinguish the cases.

Regarding duplications I see one place that can be optimised, will do